staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/fairqueuing.go

```diff
@@ -0,0 +1,621 @@
+		EnqueueTime:    qs.clk.Now(),
+		Queue:          queue,
+	}
+	if ok := qs.rejectOrEnqueue(pkt); !ok {
+		return nil
+	}
+	metrics.ObserveQueueLength(qs.name, len(queue.Requests))
+	return pkt
+
+}
+
+// removeTimedOutRequestsFromQueue rejects old requests that have been enqueued
+// past the requestWaitLimit
+func (qs *queueSetImpl) removeTimedOutRequestsFromQueue(queue *Queue) {
+	timeoutIdx := -1
+	now := qs.clk.Now()
+	pkts := queue.Requests
+	// pkts are sorted oldest -> newest
+	// can short circuit loop (break) if oldest packets are not timing out
+	// as newer packets also will not have timed out
+
+	// now - requestWaitLimit = waitLimit
+	waitLimit := now.Add(-qs.requestWaitLimit)
+	for i, pkt := range pkts {
+		if waitLimit.After(pkt.EnqueueTime) {
+			qs.wg.Increment()
+			pkt.DequeueChannel <- false
+			close(pkt.DequeueChannel)
+			// // TODO(aaron-prindle) verify this makes sense here
+			// get idx for timed out packets
+			timeoutIdx = i
+
+		} else {
+			break
+		}
+	}
+	// remove timed out packets from queue
+	if timeoutIdx != -1 {
+		// timeoutIdx + 1 to remove the last timeout pkt
+		removeIdx := timeoutIdx + 1
+
+		// remove all the timeout packets
+		queue.Requests = pkts[removeIdx:]
+		// decrement the # of requestsEnqueued
+		qs.numRequestsEnqueued -= removeIdx
+	}
+}
+
+// GetRequestsExecuting gets the # of requests which are "executing":
+// this is the# of requests/packets which have been dequeued but have not had
+// finished (via the FinishRequest method invoked after service)
+func (qs *queueSetImpl) GetRequestsExecuting() int {
+	total := 0
+	for _, queue := range qs.queues {
+		total += queue.RequestsExecuting
+	}
+	return total
+}
+
+// ChooseQueueIdx uses shuffle sharding to select an queue index
+// using a 'hashValue'.  The 'hashValue' derives a hand from a set range of
+// indexes (range 'desiredNumQueues') and returns the queue with the least queued packets
+// from a dealt hand (of size 'handSize')
+func (qs *queueSetImpl) ChooseQueueIdx(hashValue uint64, handSize int) int {
+	// TODO(aaron-prindle) currently a lock is held for this in a larger anonymous function
+	// verify that makes sense...
+
+	bestQueueIdx := -1
+	bestQueueLen := int(math.MaxInt32)
+	// desiredNumQueues is used here instead of numQueues to omit quiesce queues
+	shufflesharding.ShuffleAndDeal(hashValue, qs.desiredNumQueues, handSize, func(queueIdx int) {
+		thisLen := len(qs.queues[queueIdx].Requests)
+		if thisLen < bestQueueLen {
+			bestQueueIdx, bestQueueLen = queueIdx, thisLen
+		}
+	})
+	return bestQueueIdx
+}
+
+// rejectOrEnqueue rejects or enqueues the newly arrived request if
+// resource criteria isn't met
+func (qs *queueSetImpl) rejectOrEnqueue(packet *Request) bool {
+	queue := packet.Queue
+	curQueueLength := len(queue.Requests)
+	// rejects the newly arrived request if resource criteria not met
+	if qs.GetRequestsExecuting() >= qs.concurrencyLimit &&
+		curQueueLength >= qs.queueLengthLimit {
```

Comment:
The code says `&&` and I think that is correct.  A request has to be rejected only if it can NEITHER be launched immediately NOR queued; NOT (X OR Y) == (NOT X) AND (NOT Y).

Reply:
Got it.